src/frontend/Preprocessor.ml

```diff
@@ -12,6 +13,34 @@ let dup_exists l =
 let include_stack = Stack.create ()
 let include_paths : string list ref = ref []
 let included_files : string list ref = ref []
+let size () = Stack.length include_stack
+let init buf = Stack.push include_stack buf
+
+let current_buffer () =
+  let buf = Stack.top_exn include_stack in
+  buf
+
+let pop_buffer () = Stack.pop_exn include_stack
+
+let update_start_pos pos =
+  if size () > 1 then
+    let lexbuf = List.nth_exn (Stack.to_list include_stack) 1 in
+    lexbuf.lex_start_p <- pos
```

Comment:
I think you technically may be correct, if a user had a very long string of includes in a row.

The reason this is necessary at all is because of the way we do preprocessing. Basically, the flow of data is like this:

1. The parser is called and given the top of the preprocessor stack
2. It grabs uses that lexbuf to get one token
3. Goto 1

The issue is that inside bullet point 2 we can change the preprocessor stack, but the parser has already bound a reference to the position information of the buffer it was last called with. 

You're right in that 2 could actually add an arbitrary number of includes to the stack, so we should probably update all the start positions of them.